### PR TITLE
LUN-233: Create component for unlock exercises

### PR DESCRIPTION
**Pull request by bahaaTuffaha**
This pull request belongs to an issue on our [LUN-233](https://issues.tuerantuer.org/browse/LUN-233)

### CODE REVIEW

File: ios/Podfile.lock
Comment (SteffiStuffel) [inline ios/Podfile.lock:474]: Why these changes. Are they fine to commit @f1sh1918

File: package.json
Comment (SteffiStuffel) [inline package.json:32]: please undo this change.

File: src/routes/ExercisesScreens.tsx
Comment (SteffiStuffel) [inline src/routes/ExercisesScreens.tsx:24]: Just add a margin for left & right and remove the width.

File: src/routes/ExercisesScreens.tsx
Comment (SteffiStuffel) [inline src/routes/ExercisesScreens.tsx:23]: These predefined texts - like `HeadingText` are used, so you do not need to set font-size and font-weight everywhere. So please have a look at `src/component/text` and use the one that is appropriate for your usecase, in this case you should use `ContentTextLight` and then you can remove font-size and font-weight

File: src/routes/ExercisesScreens.tsx
Comment (SteffiStuffel) [inline src/routes/ExercisesScreens.tsx:30]: For this please add a new element to `Content.tsx` which is called `ContentTextBold` and use it.

File: src/routes/ExercisesScreens.tsx
Comment (SteffiStuffel) [inline src/routes/ExercisesScreens.tsx:48]: naming of variables is always in camelCase, so the first letter should be lower case.

File: src/routes/ExercisesScreens.tsx
Comment (SteffiStuffel) [inline src/routes/ExercisesScreens.tsx:78]: you cannot just write `prerequisite` here, it should show the name of the exercise that needs to be finished. You can determine this with your `CurrentLevel` variable.

File: src/routes/ExercisesScreens.tsx
Comment (SteffiStuffel) [inline src/routes/ExercisesScreens.tsx:73]: As the app is in german, we use the `src/constants/labels.json`. Please create entries there, which you use here. It is fine to put the english wording in the labels.json. 
Please put all text that is visible to the user into the labels.json-file.
I will later adjust them to german.

File: src/routes/exercise-finished/ExerciseFinishedScreen.tsx
Comment (SteffiStuffel) [inline src/routes/exercise-finished/ExerciseFinishedScreen.tsx:22]: Please also use camelCase here for naming.
Please also link the logic issue here, just add `LUN-131` to the comment, also please do above in ExercisesScreen in the comment.

File: src/routes/exercise-finished/ExerciseFinishedScreen.tsx
Comment (SteffiStuffel) [inline src/routes/exercise-finished/ExerciseFinishedScreen.tsx:31]: You move `props =>` to the beginning, makes it a little simler: `${prop => (ExerciseUnlocked ? prop.theme.colors.correct : prop.theme.colors.primary)};`

Same below for message.

File: src/routes/ExercisesScreens.tsx
Comment (SteffiStuffel) [inline src/routes/ExercisesScreens.tsx:76]: I think this is the wrong way round.
The `Ok, got it` button should close the modal.
The `Unlock Exercise` button should start the last exercise and therefore trigger a navigation. This can also be done with you `currentLevel` variable.

Also no need to caps-lock in the wording, as our button-component, makes the content of a button caps-lock automatically.

File: src/routes/exercise-finished/ExerciseFinishedScreen.tsx
Comment (SteffiStuffel) [inline src/routes/exercise-finished/ExerciseFinishedScreen.tsx:48]: This will only work when the device hast the correct size, but we need a solution that will also work on bigger and smaller screens. So just use `top: ??px` and `right: ??px`. Also you need to add `overflow: hidden` on the `UpperSection`. Otherwise the calculation of `right` will not work.

File: src/routes/exercise-finished/ExerciseFinishedScreen.tsx
Comment (SteffiStuffel) [inline src/routes/exercise-finished/ExerciseFinishedScreen.tsx:50]: No need for wigth and height here, as they are passed directly in the rendering below. Please remove.

**Review by SteffiStuffel — CHANGES_REQUESTED** [review]
Please also add tests with react-testing-library. E.g. a test for confirmation modal which checks that passed "children" get rendered correctly.
Also please rename the PR so it starts with `LUN-233: ` and upper case letter.

Tested on android. Looks good :tada: 
Really nice first PR!

File: src/components/ConfirmationModal.tsx
Comment (klinzo) [inline src/components/ConfirmationModal.tsx:90]: Fine for now but in the future we should probably generalize this. The only difference right now is the order of the buttons, right?

File: src/routes/exercise-finished/ExerciseFinishedScreen.tsx
Comment (klinzo) [inline src/routes/exercise-finished/ExerciseFinishedScreen.tsx:22]:
```suggestion
const exerciseUnlocked = false // TODO: LUN-131 logic
```

File: src/routes/exercise-finished/ExerciseFinishedScreen.tsx
Comment (klinzo) [inline src/routes/exercise-finished/ExerciseFinishedScreen.tsx:28]: Should be passed as prop rather than accessing global constants:
```suggestion
const UpperSection = styled.View<{ exerciseUnlocked: boolean }>`
```

File: src/routes/exercise-finished/ExerciseFinishedScreen.tsx
Comment (klinzo) [inline src/routes/exercise-finished/ExerciseFinishedScreen.tsx:37]: Why is this needed?

File: src/routes/exercise-finished/ExerciseFinishedScreen.tsx
Comment (klinzo) [inline src/routes/exercise-finished/ExerciseFinishedScreen.tsx:43]:
```suggestion
const Message = styled(HeadingBackground)<{ exerciseUnlocked: boolean }>`
```

File: src/routes/exercise-finished/ExerciseFinishedScreen.tsx
Comment (klinzo) [inline src/routes/exercise-finished/ExerciseFinishedScreen.tsx:44]:
```suggestion
  color: ${prop => (exerciseUnlocked ? prop.theme.colors.primary : prop.theme.colors.background)};
```

**Review by klinzo — CHANGES_REQUESTED** [review]
Why are there so many changes in `yarn.lock`? Should probably be reverted.
It would be good to keep your branch regularly in sync with main, so maybe you could merge that in once?

Thanks for your first PR, nice work! :tada:

File: assets/images/close-icon-white.svg
Comment (bahaaTuffaha) [reply assets/images/close-icon-white.svg:1]: Its not included in the design ,so I changed (close-circle-icon.svg) color to white..
should i consider removing it?

Comment (klinzo) [reply assets/images/close-icon-white.svg:1]: No, no need to remove it. I just meant that at least in the github preview it is not quadratic but quite tall. Not sure if thats a problem though.

Comment (klinzo) [reply assets/images/close-icon-white.svg:1]: Nvm, looks fine now.

File: src/routes/ExercisesScreens.tsx
Comment (SteffiStuffel) [reply src/routes/ExercisesScreens.tsx:76]: @bahaaTuffaha have a look at my changes, I fixed it there. Not sure why you undid it?
https://github.com/digitalfabrik/lunes-app/pull/267/commits/44f11b5b8419ac59d4ba90b5cbea780b5e2aef34

Comment (bahaaTuffaha) [reply src/routes/ExercisesScreens.tsx:76]: I didn't pull I think before modifying😅...
(EXERCISES[currentLevel].title) should i leave it as it is or the one before the current index? 
(EXERCISES[currentLevel-1].title)

Comment (SteffiStuffel) [reply src/routes/ExercisesScreens.tsx]: (EXERCISES[currentLevel-1].title)

Comment (klinzo) [reply src/routes/ExercisesScreens.tsx]: Why `- 1`? I think the currentLevel should be 0 indexed, shouldn't it?

File: ios/Podfile.lock
Comment (SteffiStuffel) [reply ios/Podfile.lock:474]: @bahaaTuffaha Please remove these changes.

File: src/components/ConfirmationModal.tsx
Comment (SteffiStuffel) [reply src/components/ConfirmationModal.tsx:90]: @klinzo should I /can you open a ticket for it?

File: src/components/text/Content.tsx
Comment (SteffiStuffel) [inline src/components/text/Content.tsx:37]: font-size should be removed as the inherited font-size from `Content` is the same.

**Review by SteffiStuffel — CHANGES_REQUESTED** [review]
- Please add tests (with react native testing)

- on the exercise screen the arrow of the locked exercises should be grey not black, see designs: https://app.zeplin.io/project/5fc7c98a435298126bcfd4ee/screen/613db6c9af77b0264444867f

- wrong level is displayed in the modal: It shows "Level 2 has to be done", but it should say "Level 1 has to be done". Also the wrong exercise is started.

- Text on the exercise finished screen is wrong. Have a look at the designs: https://app.zeplin.io/project/5fc7c98a435298126bcfd4ee/screen/613db8ae6a4311283b834885
German translation for the labels.json file is "Glückwunsch, deine nächste Übung ist jetzt freigeschaltet."

File: src/components/ConfirmationModal.tsx
Comment (klinzo) [reply src/components/ConfirmationModal.tsx:90]: Sure: https://issues.tuerantuer.org/browse/LUN-312

File: src/components/__tests__/ConfirmationModal.spec.tsx
Comment (SteffiStuffel) [inline src/components/__tests__/ConfirmationModal.spec.tsx:7]: (optional): Would be better to move this down (before renderLockingModal-function)

File: src/components/__tests__/ConfirmationModal.spec.tsx
Comment (SteffiStuffel) [inline src/components/__tests__/ConfirmationModal.spec.tsx:50]: Please add newlines between tests.

File: src/components/__tests__/ListItem.spec.tsx
Comment (SteffiStuffel) [inline src/components/__tests__/ListItem.spec.tsx:17]: Remove comment

File: src/components/__tests__/ListItem.spec.tsx
Comment (SteffiStuffel) [inline src/components/__tests__/ListItem.spec.tsx:19]: please rename to arrowDisabled, also you can define it to be false if no parameter is passed by adding defining it like this `const renderDisciplineItem = (arrowDisabled = false): RenderAPI =>` and then where it is false you can just call it without a parameter.

File: src/components/__tests__/ListItem.spec.tsx
Comment (SteffiStuffel) [inline src/components/__tests__/ListItem.spec.tsx:48]: This test is not testing it the press is correctly handled. So please rename it to "should show arrow disabled", and also you can remove everything from this test except the first three lines, as it does not make sense to 1) Check for title color, as this does not depend on the arrowDisabling (what is tested by this test) and 2) not trigger an action and then check for the same stuff again.

File: src/components/__tests__/ListItem.spec.tsx
Comment (SteffiStuffel) [inline src/components/__tests__/ListItem.spec.tsx:82]: Please remove this test, it does not test anything at all, which is not tested by other tests in this file.

File: src/components/__tests__/ListItem.spec.tsx
Comment (SteffiStuffel) [inline src/components/__tests__/ListItem.spec.tsx:116]: Please remove this test, it also does not test anything at all, which is not tested by other tests in this file.

File: src/components/text/Content.tsx
Comment (SteffiStuffel) [inline src/components/text/Content.tsx]: Please add this again. Chaning the components in `components/text/...` will change the text everywhere in the app. By my last comment, because of which you removed this, I meant you should remove the `font-size`-Attribute from your styled component that you have created, as if you use

```
const MyComponent = styled(Conent)'
 font-size = prop.theme.fonts.defaultFontSize
'
```

the font-size here is just not needed as `MyComponent` just extends `Content` and therefore all attributes of Content (like the font-size) are already applied to MyComponent

File: src/routes/ExercisesScreens.tsx
Comment (SteffiStuffel) [reply src/routes/ExercisesScreens.tsx:40]: No I think it is fine with `<=`. Current level is the level the user is currently on, so if finishing level 2, the currentLevel is 2, so it would be valid the user can start the exercise of level 2 or am I missing something here?

**Review by SteffiStuffel — COMMENTED** [review]
A test for ExercisesScreen would be nice, which handels if the correct action is triggered on pressing (is modal show or is navigation triggered), but if you struggle with this you can skip it, as it would be really nice to get you PR merged soon.

Also please fix linting and prettier errors.

File: src/components/text/Content.tsx
Comment (bahaaTuffaha) [reply src/components/text/Content.tsx]: Oh .. my bad 😥

File: .eslintrc.js
Comment (klinzo) [inline .eslintrc.js:58]: Please revert

File: ios/Podfile.lock
Comment (klinzo) [inline ios/Podfile.lock:632]: Please revert

**Review by klinzo — APPROVED** [review]
Approved to not block merging this, did not review/test again.

File: src/components/text/Content.tsx
Comment (SteffiStuffel) [reply src/components/text/Content.tsx:37]: @bahaaTuffaha please adjust this.

**Review by SteffiStuffel — COMMENTED** [review]
Please have a look at the hidden conversations of this PR, some are not adressed yet.

**Comment by klinzo** [comment]
@bahaaTuffaha just a general hint: Please resolve or respond to comments you have addressed in future, otherwise it is really hard to tell what you did do, did not do intentionally and what comments you just overlooked by mistake. You don't have to do this here anymore though.

File: src/components/__tests__/ListItem.spec.tsx
Comment (bahaaTuffaha) [reply src/components/__tests__/ListItem.spec.tsx]: I have removed this test already..
should I try writing another test?

Comment (SteffiStuffel) [reply src/components/__tests__/ListItem.spec.tsx]: no its fine, like it is.

File: src/components/text/Content.tsx
Comment (SteffiStuffel) [reply src/components/text/Content.tsx:37]: as discussed, it is fine as it is

**Review by SteffiStuffel — APPROVED** [review]
Nice :tada: 

Tested on Android, looks really good. You can merge this now, after you resolved the conflicts.

**Pull request merged**
